doc/udev-rules.org

```diff
@@ -0,0 +1,244 @@
+* Udev Rules in yast-network
+
+https://github.com/yast/yast-network
+
+** What are Udev Rules
+
+If a computer has two network cards, how does it determine which one is *eth0*
+and which one is *eth1*?
+
+A Linux machine may have several network interfaces, such as eth0, eth1,
+wlan0, enp0s3.
+These udev rules ensure the correct assignment of interface names.
+
+_(
+  "<p><b>Udev Rules</b> are rules for the kernel device manager that allow\n" \
+    "associating the MAC address or BusID of the network device with its name (for\n" \
+    "example, eth1, wlan0 ) and assures a persistent device name upon reboot.\n"
+)
+
+For a NIC a widget exists where you can change its name and choose whether it
+will be pinned to its MAC (Ethernet address), or its PCI bus address.
+
+** Why
+
+The code of yast-network is messy, so Udev Rules is one area to clean it up.
+It is simple enough and well defined.
+
+** Plan
+
+- [x] collect all mentions of "udev" in yast-network
+- [ ] read them all and design a decent API
+- [ ] cover the code with tests
+- [ ] use the new API, possibly using old code underneath
+- [ ] replace old implementation with a better one
+
+** AY interface
+
+The rnc schema says all elements are optional
+but the SLE12(also 15) docs says name+rule+value are all required
+
+https://www.suse.com/documentation/sles-12/singlehtml/book_autoyast/book_autoyast.html#CreateProfile.Network.names
+
+Semantics: the device *name* is identified
+by the field with the key *rule* (sic!) having the value *value*
+(that does not allow some complex identifications)
+
+#+BEGIN_SRC xml
+<networking>
+  <net-udev config:type="list">
+    <rule>
+      <name>eth0</name>
+      <rule>ATTR{address}</rule>
+      <value>00:30:6E:08:FF:80</value>
+    </rule>
+  </net-udev>
+</networking>
+#+END_SRC
+
+** where is "udev" mentioned?
```

Comment:
is it used also outside of yast2-network? it is important info, especially to collect requirements.

Reply:
"Udev" is mentioned a lot in yast-storage-ng, but the situation is quite different:

What network and storage devices have in common is that both kinds of devices have a *bus position*, and some kind of *unique id*, and a *short kernel name*.

Storage devices have a device file that exists in the filesystem, so naturally you can make symlinks pointing to `/dev/sda1` from `/dev/disk/by-path/pci-0000:00:1f.2-scsi-0:0:0:0-part1` or from `/dev/disk/by-id/ata-ST3500418AS_9VMN8X8L-part1`. For which there are udev rules that simply *work* and YaST does not have to care.

Networking devices **suck** because they **do not exist** in the filesystem, so we cannot have multiple symlinks to a "canonical" device name. We must embed the paths and ids in the udev rules, even to choose one of the naming schemes over the others (name NICs by MAC, not by PCI bus position).

Reply:
**yast-country** has one place which needs to write a simple udev rule, which is done by treating it as a string. I think it is better to do it so without adding a udev library dependency:

https://github.com/yast/yast-country/blob/1538404d23c64c5733862abd0ded363f47594732/keyboard/src/modules/Keyboard.rb#L1340